[RunAllTests] Fix #2941: Fix glide image loading in Bazel

BenHenning opened the pull request:
Fix #2941

This properly hooks up Glide's annotation processor so that RepositoryGlideModule is actually called to initialize the Glide module. Without this, images will not load correctly on Bazel builds unless they utilize local drawables.

This builds off of changes introduced in #3025 to consistently hook up the Glide compiler plugin. It essentially behaves the same way except we're more limited in what we can export with regards to Glide-related dependencies (see the disclaimer in code).

Note that there isn't a clear way to test this with automation, but if you want to verify it locally:
- Check out the branch
- On your device, select Oppia as the debug app & enable 'wait for debugger'
- Build & install the app: ``bazel build //:oppia && adb install -r bazel-bin/oppia.apk``
- Point a breakpoint in the beginning of ``RepositoryGlideModule.registerComponents`` via Android Studio
- Attach the debugger from Android Studio
- Verify that the line is hit shortly after the splash screen (it won't be if you tried the same repro steps on ``develop``

I think we'll eventually be able to test this with screenshot-enabled end-to-end tests, but those aren't even on the roadmap at the moment. :) Humans will likely need to verify this doesn't regress (fortunately it's very obvious--no Glide images will load at all).

BenHenning committed:
Add support for loading proto versions of lessons.

BenHenning committed:
Upgrade Espresso versions to work around build issues.

Note that I can't actually verify Espresso tests are working with these
versions since the Bazel project doesn't yet have Espresso tests set up.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Lint fix post-copying from #2927.

BenHenning committed:
Fix Glide in Bazel.

BenHenning committed:
Remove image loading annotation.

This will be added in a later branch in the chain.

BenHenning committed:
Add caching module to needed test suites.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning commented:
/cc @fsharpasharp

BenHenning commented:
/cc @anandwana001

anandwana001 commented on tools/BUILD.bazel:
```diff
@@ -0,0 +1,15 @@
+load("@rules_jvm_external//:defs.bzl", "artifact")
```
looks like we are not using `artifact`!!

BenHenning replied:
Good find--removed.

anandwana001 commented on tools/BUILD.bazel:
```diff
@@ -0,0 +1,15 @@
+load("@rules_jvm_external//:defs.bzl", "artifact")
+
+java_plugin(
```
for java plugin, we should use plugin load.
`load("@rules_java//java:defs.bzl", "java_plugin")`

BenHenning replied:
Fixed.

anandwana001 reviewed (changes requested):
Let's see once we merge blocking pr - 2978

BenHenning committed:
Add support for importing text protos as assets.

This actually provides Bazel-only support for converting text protos to
binary and including those as assets so that we can avoid checking in
binary protos to the codebase.

BenHenning committed:
Add gitignore for Android Studio Bazel plugin output.

BenHenning committed:
Import textproto versions of existing dev assets.

This imports local conversions of all existing dev assets as text proto
leveraging the Bazel-only conversion system to convert these to binary
protos and include them as assets.

This is a necessary step to both eventually replace JSON assets with
text proto, and to test that the new proto loading pipeline is working
as designed.

BenHenning committed:
Undo Espresso version change since it breaks Bazel tests.

BenHenning committed:
Fix tests.

Various attempts to fix existing tests + move some of the domain tests
over to actually using proto when running with Bazel.

App module tests are hanging currently for unknown reasons (only in
Bazel).

BenHenning committed:
Use correct Espresso core.

3.1.0 causes tests to hang indefinitely when run with Bazel.

BenHenning committed:
Move coroutine dispatchers.

This moves the coroutine dispatchers to a new threading subpackage. It
also simplifies the TestCoroutineDispatcher interface & implementations
(which required moving runUntilIdle to CoroutineExecutorServiceTest).

The test coroutine annotations were also moved to their own files.

BenHenning committed:
Refactor testing utilities build graph.

This introduces new Bazel libraries for:
- Test dispatchers
- Fake system/Oppia clocks
- Robolectric dependencies

BenHenning committed:
Introduce dedicated test for TestCoroutineDispatcher.

This also fixes some threading issues in the dispatcher, and clarifies
some of its API that was previously unclear (and can lead to subtle race
conditions).

BenHenning committed:
Add tests for TestCoroutineDispatcher.

This required a bunch of refactoring, adding the first examples of
dedicated Bazel test build files, and introducing a new test pattern for
sharing code between different implementations. Further, it refined the
API for TesCoroutineDispatcher & fixed some issues in the
implementations (especially Robolectric).

Tests verified as non-flaky in Robolectric across 1000 runs, and 0.2%
flaky across 1000 runs for Espresso (though that may be fixed; waiting
on a follow-up run to confirm stability).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Workaround ktlint semicolon issue.

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix test post-merge.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Move CoroutineDispatcher to be part of threading.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into fix-protobuf-duplication-issue

BenHenning committed:
Undo version upgrade.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	.gitignore
	app/src/sharedTest/java/org/oppia/android/app/story/StoryActivityTest.kt

BenHenning committed:
Fix broken tests.

BenHenning committed:
Lint fixes.

BenHenning committed:
Add a bit more proto loading test coverage.

BenHenning committed:
Comment fix.

BenHenning committed:
Lint fixes.

BenHenning committed:
Address earlier TODO.

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	tools/BUILD.bazel

BenHenning committed:
Post-merge restructure.

BenHenning commented:
Setting '[RunAllTests]' per https://github.com/oppia/oppia-android/pull/2978#issuecomment-818309746.

BenHenning commented:
Thanks @anandwana001. This is now ready for full review. Please do not block approval on #2978 unless you have a very particular reason to (in which case, please reach out). There's a long chain of PRs and it will be very time consuming to merge them in one-at-a-time if each is blocked only on the previous PR for approval.

Edit: as a follow-up, I'll also be merging #2832 into this PR now that it's in develop. That PR will prevent me from being able to merge into the base branch since it isn't develop.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

Conflicts:
	data/BUILD.bazel

BenHenning committed:
Post-merge fixes.

This removes unneeded dependencies from PersistentCacheStore, including
Glide.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Fix broken test build.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into fix-glide-image-loading

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Post-merge lint fixes.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

anandwana001 commented:
```
ERROR: /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/external/maven/BUILD:3353:11: Desugaring external/maven/v1/https/repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-reflect/1.4.10/kotlin-reflect-1.4.10.jar for Android failed (Exit 1)
java.lang.IllegalStateException
        at com.google.common.base.Preconditions.checkState(Preconditions.java:491)
        at com.google.devtools.build.android.desugar.Desugar.desugarClassesInInput(Desugar.java:546)
        at com.google.devtools.build.android.desugar.Desugar.desugarOneInput(Desugar.java:320)
        at com.google.devtools.build.android.desugar.Desugar.desugar(Desugar.java:240)
        at com.google.devtools.build.android.desugar.Desugar.processRequest(Desugar.java:1014)
        at com.google.devtools.build.android.desugar.Desugar.runPersistentWorker(Desugar.java:974)
        at com.google.devtools.build.android.desugar.Desugar.main(Desugar.java:952)
Target //:oppia failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1015.784s, Critical Path: 103.44s
INFO: 1198 processes: 1094 darwin-sandbox, 104 worker.
FAILED: Build did NOT complete successfully
```

Running the given bazel command in description and getting this error. 
@fsharpasharp are you also facing the similar issue in checking this pr?

fsharpasharp commented:
> ```
> ERROR: /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/external/maven/BUILD:3353:11: Desugaring external/maven/v1/https/repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-reflect/1.4.10/kotlin-reflect-1.4.10.jar for Android failed (Exit 1)
> java.lang.IllegalStateException
>         at com.google.common.base.Preconditions.checkState(Preconditions.java:491)
>         at com.google.devtools.build.android.desugar.Desugar.desugarClassesInInput(Desugar.java:546)
>         at com.google.devtools.build.android.desugar.Desugar.desugarOneInput(Desugar.java:320)
>         at com.google.devtools.build.android.desugar.Desugar.desugar(Desugar.java:240)
>         at com.google.devtools.build.android.desugar.Desugar.processRequest(Desugar.java:1014)
>         at com.google.devtools.build.android.desugar.Desugar.runPersistentWorker(Desugar.java:974)
>         at com.google.devtools.build.android.desugar.Desugar.main(Desugar.java:952)
> Target //:oppia failed to build
> Use --verbose_failures to see the command lines of failed build steps.
> INFO: Elapsed time: 1015.784s, Critical Path: 103.44s
> INFO: 1198 processes: 1094 darwin-sandbox, 104 worker.
> FAILED: Build did NOT complete successfully
> ```
> 
> Running the given bazel command in description and getting this error.
> @fsharpasharp are you also facing the similar issue in checking this pr?

I'll take a look today.

Edit: I got

```
ERROR: /home/jan/oppia-android/app/BUILD.bazel:555:19: Building app/libview_models_base.jar (1 source file, 1 source jar) and running annotation processors (ComponentProcessor, GlideAnnotationProcessor, ProcessDataBinding) failed: (Exit 1): java failed: error executing command external/remotejdk11_linux/bin/java -XX:+UseParallelOldGC -XX:-CompactStrings '--patch-module=java.compiler=external/remote_java_tools_linux/java_tools/java_compiler.jar' ... (remaining 15 argument(s) skipped)
bazel-out/android-armeabi-v7a-fastbuild/bin/app/databinding/view_models_base/DataBindingInfo.java:3: error: cannot find symbol
import android.databinding.BindingBuildInfo;
                          ^
  symbol:   class BindingBuildInfo
  location: package android.databinding
bazel-out/android-armeabi-v7a-fastbuild/bin/app/databinding/view_models_base/DataBindingInfo.java:10: error: cannot find symbol
@BindingBuildInfo(
 ^
  symbol: class BindingBuildInfo
Target //:oppia failed to build
```

BenHenning commented:
@anandwana001 can you try a clean & rebuild? It's building fine for both me & CI, and I've actually never seen that desugar error before. I'm wondering if it's a transient issue.

@fsharpasharp do you have the instructions at https://github.com/oppia/oppia-bazel-tools setup? That includes adding ``build --android_databinding_use_androidx`` to your ~/.bazelrc file which I would expect to resolve the error you linked (the giveaway is that it's trying to find android.databinding.BindingBuildInfo not androidx.databinding.BindingBuildInfo).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'add-support-for-proto-lesson-loading-from-local-assets' into fix-glide-image-loading

BenHenning commented:
Going to test and see if we can avoid the ``[RunAllTests]`` now that #2978 is merged. That would definitely speed up CI for this & downstream PRs that aren't as broadly affecting the build graph or app.

BenHenning commented:
Closing & reopening PR to retrigger full CI.

BenHenning commented:
I think I was too optimistic--Bazel file changes in this PR seems to break compute_affected_tests, too. :(

BenHenning commented:
Re-triggering workflow with full test suite.

BenHenning commented:
@anandwana001 PTAL. This is blocking the entire alpha PR chain.

anandwana001 reviewed (approved):
LGTM, approving the PR as don't want to block until my Bazel issue get's solved. 
I think there is some missing part of Bazel on my side due to a poor network connection. I will try to setup again today.

anandwana001 commented:
Error ->

```
/oppia/oppia-android/model/BUILD.bazel:52:14: Generating JavaLite proto_library //model:event_logger_proto failed: (Aborted): sandbox-exec failed: error executing command 
  (cd /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/sandbox/darwin-sandbox/1583/execroot/__main__ && \
  exec env - \
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/akshaynandwana/Library/Python/2.7/bin \
    TMPDIR=/var/folders/gp/h5bq28910jlb0zynnt11n0kc0000gn/T/ \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/sandbox/darwin-sandbox/1583/sandbox.sb /var/tmp/_bazel_akshaynandwana/install/5ccdfa437f9a94a683a2b6383c9f8794/process-wrapper '--timeout=0' '--kill_delay=15' bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/protoc '--java_out=lite:bazel-out/android-armeabi-v7a-fastbuild/bin/model/event_logger_proto-lite-src.jar' '-Imodel/src/main/proto/oppia_logger.proto=model/src/main/proto/oppia_logger.proto' --direct_dependencies model/src/main/proto/oppia_logger.proto '--direct_dependencies_violation_msg=%s is imported, but //model:event_logger_proto doesn'\''t directly depend on a proto_library that '\''srcs'\'' it.' model/src/main/proto/oppia_logger.proto) sandbox-exec failed: error executing command 
  (cd /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/sandbox/darwin-sandbox/1583/execroot/__main__ && \
  exec env - \
    PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/akshaynandwana/Library/Python/2.7/bin \
    TMPDIR=/var/folders/gp/h5bq28910jlb0zynnt11n0kc0000gn/T/ \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/sandbox/darwin-sandbox/1583/sandbox.sb /var/tmp/_bazel_akshaynandwana/install/5ccdfa437f9a94a683a2b6383c9f8794/process-wrapper '--timeout=0' '--kill_delay=15' bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/protoc '--java_out=lite:bazel-out/android-armeabi-v7a-fastbuild/bin/model/event_logger_proto-lite-src.jar' '-Imodel/src/main/proto/oppia_logger.proto=model/src/main/proto/oppia_logger.proto' --direct_dependencies model/src/main/proto/oppia_logger.proto '--direct_dependencies_violation_msg=%s is imported, but //model:event_logger_proto doesn'\''t directly depend on a proto_library that '\''srcs'\'' it.' model/src/main/proto/oppia_logger.proto)
dyld: Symbol not found: __ZNK6google8protobuf8compiler3php9Generator11GenerateAllERKNSt3__16vectorIPKNS0_14FileDescriptorENS4_9allocatorIS8_EEEERKNS4_12basic_stringIcNS4_11char_traitsIcEENS9_IcEEEEPNS1_16GeneratorContextEPSI_
  Referenced from: /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/sandbox/darwin-sandbox/1583/execroot/__main__/bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/protoc
  Expected in: flat namespace
 in /private/var/tmp/_bazel_akshaynandwana/0e0a7569f455d331a530cc35527abf40/sandbox/darwin-sandbox/1583/execroot/__main__/bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/protoc
Target //:oppia failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 316.057s, Critical Path: 53.48s
INFO: 420 processes: 31 internal, 387 darwin-sandbox, 2 worker.
FAILED: Build did NOT complete successfully
```

BenHenning commented:
Thanks @anandwana001. Let me know if you want more help investigating that issue. The build seems to work for me & CI, so there might be an environment disparity issue to resolve on your end.
